sql/delegated_staking.sql

```diff
@@ -0,0 +1,71 @@
+--
+--We want to be able to show delegated staking creation, update, and withdrawal events in the DAG explorer so we’ll need support for them through the API.
+--
+--Add indexing for:
+--
+--DelegatedStake (Create/Update can be combined here)
+--
+--DelegatedStakeWithdrawal
+--
+--Add endpoints:
+--
+--/delegated-stakes
+--
+--/delegated-stakes/:hash
+--
+--/addresses/:address/delegated-stakes
+--
+--/delegated-stake-withdrawals
+--
+--/delegated-stake-withdrawals/:hash
+--
+--/addresses/:address/delegated-stake-withdrawals
+
+CREATE TABLE delegate_stake_create_events (
+    hash varchar PRIMARY KEY,
+    ordinal int8 NOT NULL,
+    source_addr varchar NOT NULL REFERENCES addresses(address) ON DELETE CASCADE, --FK address
+    node_id varchar NOT NULL,
+    amount int8 NOT NULL,
+    fee int8 NOT NULL DEFAULT 0,
+    token_lock_ref varchar NOT NULL REFERENCES dag_token_locks(hash) ON DELETE CASCADE, --FK token lock
+    parent_hash varchar NOT NULL,
+    global_snapshot_hash varchar NOT NULL REFERENCES global_snapshots(hash) ON DELETE CASCADE
+);
+CREATE INDEX delegate_stake_create_events_source_addr_idx ON public.delegate_stake_create_events USING btree (source_addr);
+CREATE INDEX delegate_stake_create_events_token_lock_ref_idx ON public.delegate_stake_create_events USING btree (token_lock_ref);
+CREATE INDEX delegate_stake_create_events_global_snapshot_hash_idx ON public.delegate_stake_create_events USING btree (global_snapshot_hash);
+
+
+
+CREATE TABLE delegate_stake_withdraw_events (
+    hash varchar PRIMARY KEY,
+    source_addr varchar NOT NULL REFERENCES addresses(address) ON DELETE CASCADE,
+    stake_ref varchar NOT NULL REFERENCES delegate_stake_create_events(hash) ON DELETE CASCADE,
+    global_snapshot_hash varchar NOT NULL REFERENCES global_snapshots(hash) ON DELETE CASCADE
+);
+CREATE INDEX delegate_stake_withdraw_events_source_addr_idx ON public.delegate_stake_withdraw_events USING btree (source_addr);
+CREATE INDEX delegate_stake_withdraw_events_stake_ref_idx ON public.delegate_stake_withdraw_events USING btree (stake_ref);
+CREATE INDEX delegate_stake_withdraw_events_global_snapshot_hash_idx ON public.delegate_stake_withdraw_events USING btree (global_snapshot_hash);
+
+
+CREATE TABLE delegate_stake_balance_changes (
+	global_snapshot_hash varchar NOT NULL REFERENCES global_snapshots(hash) ON DELETE CASCADE,
+	global_snapshot_ordinal int8 NOT NULL REFERENCES global_snapshots(ordinal) ON DELETE CASCADE
+	address varchar NOT NULL REFERENCES addresses(address) ON DELETE CASCADE,
+	node_id varchar NOT NULL,
+	balance int8 NOT NULL,
+	rewards int8 NOT NULL,
+	created_at timestamp DEFAULT now() NOT NULL,
+	updated_at timestamp DEFAULT now() NOT NULL
+);
+CREATE INDEX delegate_stake_balance_changes_idx ON public.delegate_stake_balance_changes USING btree (address);
```

Comment:
NodeID will be the other way we'll need to query this so it would make sense to add an index for that here.